### DISCUSSION TITLE
Exception cause in browser console

### DISCUSSION THREAD

**Discussion by faveoled** [google/j2cl#181] [Q&A] [answered]
The code throws an exception with a cause:
```
        RuntimeException inner = new RuntimeException("Inner");
        throw new RuntimeException("Outer", inner);
```
I have the following in my browser console:
![image](https://user-images.githubusercontent.com/85760289/219007942-c42409b8-0dac-4478-b659-6bf7aa30e21e.png)
Is there a way to also have "Inner" exception message in the browser console? I expected it to appear similar to regular Java exceptions.
JRE sample output:
```
Exception in thread "main" java.lang.RuntimeException: Outer
	at MyClass.main(MyClass.java:4)
Caused by: java.lang.RuntimeException: Inner
	at MyClass.main(MyClass.java:3)
```

**Comment by gkdn** [comment, marked as answer]
We populate the `cause` property of the JS errors.

This is relatively recently [standardized](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause) so DevTools might not have caught up with them yet. You can file a bug with Chrome or the browser that you are developing with.

Being said that if you pause, then you can still inspect the error object and inspect/print the `cause` field of it.

**Reply by faveoled** [reply]
I've tried one example from your link and my browser does print a cause:
![image](https://user-images.githubusercontent.com/85760289/219317089-64f0e6d2-963d-4ac8-9b03-a8feaab01a09.png)

**Reply by faveoled** [reply]
I've stored the exception as a global and the cause is undefined:
![image](https://user-images.githubusercontent.com/85760289/219325499-7b710a29-f12d-4542-8230-b13e656f7b5d.png)

**Reply by faveoled** [reply]
My J2CL is of version `0.11.0-9336533b6`

**Reply by niloc132** [reply]
Please note that this version number, `0.11.0-9336533b6` is deployed to maven central from the github.com/vertispan/j2cl fork - we deliberately add notes in the pom that this is not the upstream version and bugs should be filed against our fork (or j2cl-maven-plugin) instead as we sometimes lag behind, and have a few custom fixes and don't want the upstream repo to get blamed for our issues.

The `9336533b6` at the end of the version number means that this is based on commit https://github.com/google/j2cl/commit/9336533b6 in the upstream repo, which is almost six months old. We will release a newer build when we update our modifications to latest, or when google pushes a tag or otherwise denotes a "stable release".

--

With that said, Throwable.getCause() and the matching "this js error has a backing java exception" wiring appears that it was [implemented in July](https://github.com/google/j2cl/commit/5f203bd5b), so should be included in the above release. I see more recent changes to those files for wasm, but I don't think they should affect this feature.

**Reply by faveoled** [reply]
> this is not the upstream version and bugs should be filed against our fork

Thank you, I wasn't aware of that

**Reply by gkdn** [reply]
Actually we started populating this property quite a while [back](https://github.com/google/j2cl/commit/1e58670b8502e691f9f41764fce526f859c23f36) but further iterations of Throwable enabled potential name mangling of `cause`  on  the JS Error object.

After this is standardized and JsCompiler added `cause` to the externs (around June 2022) this should be all working as expected for JS debugging of the optimized builds.

**Reply by niloc132** [reply]
Thanks for the added suggestion of closure-compiler - in the commit where j2cl-maven-plugin updated to 0.11.0, it also was updated to the nov 2022 release of closure compiler, see https://github.com/Vertispan/j2clmavenplugin/commit/b4159cdd612bb608175ffc5368ac329c949eaecd#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8

Assuming you are using j2cl-maven-plugin 0.21-SNAPSHOT, you likely have the dependencies required for this, or are you running with our maven release in some other way such that an out of date closure might be in use? If you are running in the j2cl-maven-plugin and still seeing this issue, I'd feel better if you [filed a ticket with us](https://github.com/Vertispan/j2clmavenplugin/issues/new) rather than here, to make sure we rule out our own forks having caused any of this.

**Reply by faveoled** [reply]
I had `j2cl-maven-plugin` `0.20`, J2CL version was configured via a property which likely wasn't even picked up. Maybe I'll try `0.21` when it's released

**Reply by niloc132** [reply]
Yes, that will have an older closure-compiler too. Perhaps try the snapshot build, we're looking to release it soon.
